src/legendhpges/icpc.py

```diff
@@ -0,0 +1,235 @@
+        # first point
+        rlist.append(wrad)
+        zlist.append(det_h - wgap)
+
+        # groove
+        groovedata = datadict["groove"]
+        grad = groovedata["outer_radius_in_mm"]
+        gdepth = groovedata["depth_in_mm"]
+        gwidth = groovedata["width_in_mm"]
+
+        # taper top
+        tapertopdata = datadict["taper"]["top"]
+        tinner = tapertopdata["inner"]
+        alpha = tinner["angle_in_deg"]
+        tinheight = tinner["height_in_mm"]
+        tstepinner = tinheight * tan(np.deg2rad(alpha))
+        if tinheight > 0:  # inner taper exists
+            rlist.append(wrad)
+            zlist.append(det_h - tinheight)
+            rlist.append(wrad + tstepinner)
+            zlist.append(0)
+        else:
+            rlist.append(wrad)
+            zlist.append(0)
+
+        touter = tapertopdata["outer"]
+        alpha = touter["angle_in_deg"]
+        toutheight = touter["height_in_mm"]
+        tstepouter = toutheight * tan(np.deg2rad(alpha))
+        if toutheight > 0:  # outer taper exists
+            rlist.append(det_r - tstepouter)
+            zlist.append(0)
+            rlist.append(det_r)
+            zlist.append(toutheight)
+        else:
+            rlist.append(det_r)
+            zlist.append(0)
+
+        taperbotdata = datadict["taper"]["bottom"]
+        touter = taperbotdata["outer"]
+        alpha = touter["angle_in_deg"]
+        toutheight = touter["height_in_mm"]
+        tstepouter = toutheight * tan(np.deg2rad(alpha))
+        if toutheight > 0:  # bottom taper exists
+            rlist.append(det_r)
+            zlist.append(det_h - toutheight)
+            rlist.append(det_r - tstepouter)
+            zlist.append(det_h)
+        else:
+            rlist.append(det_r)
+            zlist.append(det_h)
+
+        # walk rest of shape
+        rlist.append(grad)
+        zlist.append(det_h)
+        rlist.append(grad)
+        zlist.append(det_h - gdepth)
+        rlist.append(grad - gwidth)
+        zlist.append(det_h - gdepth)
+        rlist.append(grad - gwidth)
+        zlist.append(det_h)
+        rlist.append(0)
+        zlist.append(det_h)
+        rlist.append(0)
+        zlist.append(det_h - wgap)
+
+        return rlist, zlist
+
+    def _build_crystal(self, data_dict, reg, material):
+        """
+        Build the crystal from JSON.
+
+        Parameters
+        ----------
+        data_dict : dict
+                dictionary data describing crystal shape.
+
+        Returns
+        -------
+        ge_lv : pg4.geant4.LogicalVolume
+            the crystal logical volume; placement in main code.
+
+        """
+        # return ordered r,z lists, default unit [mm]
+        rlist, zlist = self._decode_polycone(data_dict)
+
+        # build generic polycone, and logical volume, default [mm]
+        ge_solid = pg4.geant4.solid.GenericPolycone("Ge", 0, 2 * pi, rlist, zlist, reg)
+        ge_lv = pg4.geant4.LogicalVolume(ge_solid, material, "GeLV", reg)
```

Comment:
What about setting G4 solid and logical volume names to `self.detname`? We might also want an optional argument in the constructor if users need to to override it.